HSEARCH-4566 SPI for "hot" restart that allows compatible schema changes with Lucene without losing data in local-heap indexes

fax4ever opened the pull request:
https://hibernate.atlassian.net/browse/HSEARCH-4566

draft => ~~I'm going to add a test~~ => done but ~~the test does not pass (still draft!)~~ => done

limited => with this contribution the promise:

> While Hibernate Search is restarting, the old Hibernate Search instance can be used.

won't be addressed

fax4ever commented:
> Kudos, SonarCloud Quality Gate passed!    [![Quality Gate passed](https://camo.githubusercontent.com/e4a90cf346b46d44c5652111d68603714bfb7693043d07a95992de3a30bfce04/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f5175616c6974794761746542616467652f7061737365642d313670782e706e67)](https://sonarcloud.io/dashboard?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=3029)
> 
> [![Bug](https://camo.githubusercontent.com/4c6102327f5a954f9c8acaf2e2714183157a9e41717b371b2cd585cf25057310/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6275672d313670782e706e67)](https://sonarcloud.io/project/issues?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=3029&resolved=false&types=BUG) [![A](https://camo.githubusercontent.com/1cba125a897d7fa47033a3b3b2be2bbee680d34d4f004a215564659b853fb201/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412d313670782e706e67)](https://sonarcloud.io/project/issues?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=3029&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=3029&resolved=false&types=BUG) [![Vulnerability](https://camo.githubusercontent.com/3ba1ee49636ffc3427e38649a9f8a65ee392f28e8a662fcf96ce24cefbb520e9/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f76756c6e65726162696c6974792d313670782e706e67)](https://sonarcloud.io/project/issues?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=3029&resolved=false&types=VULNERABILITY) [![A](https://camo.githubusercontent.com/1cba125a897d7fa47033a3b3b2be2bbee680d34d4f004a215564659b853fb201/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412d313670782e706e67)](https://sonarcloud.io/project/issues?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=3029&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=3029&resolved=false&types=VULNERABILITY) [![Security Hotspot](https://camo.githubusercontent.com/fb735cbe76f8d5e1679c76ce83b740ceb1eaf62de4f7bf88623dc9953261aff7/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f73656375726974795f686f7473706f742d313670782e706e67)](https://sonarcloud.io/project/security_hotspots?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=3029&resolved=false&types=SECURITY_HOTSPOT) [![A](https://camo.githubusercontent.com/1cba125a897d7fa47033a3b3b2be2bbee680d34d4f004a215564659b853fb201/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412d313670782e706e67)](https://sonarcloud.io/project/security_hotspots?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=3029&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=3029&resolved=false&types=SECURITY_HOTSPOT) [![Code Smell](https://camo.githubusercontent.com/8fe18b2dfb6f7d4e44582f281b29f617eb5ae07c248d2002ca586e91da219212/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f636f64655f736d656c6c2d313670782e706e67)](https://sonarcloud.io/project/issues?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=3029&resolved=false&types=CODE_SMELL) [![A](https://camo.githubusercontent.com/1cba125a897d7fa47033a3b3b2be2bbee680d34d4f004a215564659b853fb201/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412d313670782e706e67)](https://sonarcloud.io/project/issues?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=3029&resolved=false&types=CODE_SMELL) [6 Code Smells](https://sonarcloud.io/project/issues?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=3029&resolved=false&types=CODE_SMELL)
> 
> [![No Coverage information](https://camo.githubusercontent.com/86785eaa572e061f5ecb3f8bf1cbb7b8eca44925083f92d067805b4362cead0c/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f436f76657261676543686172742f4e6f436f766572616765496e666f2d313670782e706e67)](https://sonarcloud.io/component_measures?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=3029&metric=coverage&view=list) No Coverage information [![No Duplication information](https://camo.githubusercontent.com/503f02cc51df080d6036c68f91b1f9d7dc59b3243422d9f2c5549345cb2dbadd/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f4475706c69636174696f6e732f4e6f4475706c69636174696f6e496e666f2d313670782e706e67)](https://sonarcloud.io/component_measures?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=3029&metric=duplicated_lines_density&view=list) No Duplication information

addressed with my last forced push

yrodiere commented on engine/src/main/java/org/hibernate/search/engine/backend/index/spi/IndexManagerState.java:
```diff
@@ -0,0 +1,17 @@
+ * A state from an existing {@link IndexManagerImplementor}
+ * that can be used to create a new instance of it.
+ */
+public interface IndexManagerState {
```
Considering we'll potentially need to save state anywhere (backend, index manager, shard holder, shard, ...), and state will consist in potentially many types of values, I'd go with a more generic abstraction. Something like this:

```java
public class SavedState {

    private static final SavedState EMPTY = new SavedState.Builder().build();

    public static SavedState empty() {
        return EMPTY;
    }

    private final Map<Key<?>, Object> content;

    private SavedState(Builder builder) {
        this.content = builder.content;
    }

    public <T> Optional<T> get(Key<T> key, T defaultValue) {
        return Optional.ofNullable( (T) content.get( key ) );
    }

    public static <T> Key<T> key(String name) {
        Contracts.assertNotNullNorEmpty( name, "name" );
        return new Key<>( name );
    }

    public final class Key<T> {

        private final String name;

        private Key(String name) {
            this.name = name;
        }

        @Override
        public String toString() {
            return getClass().getSimpleName() + "[" + name + "]";
        }

        @Override
        public boolean equals(Object o) {
            if ( this == o ) {
                return true;
            }
            if ( o == null || getClass() != o.getClass() ) {
                return false;
            }
            Key<?> that = (Key<?>) o;
            return Objects.equals( name, that.name );
        }

        @Override
        public int hashCode() {
            return Objects.hash( name );
        }

    }

    public static Builder builder() {
        return new Builder();
    }

    public static final class Builder {
        private final Map<Key<?>, Object> content = new HashMap<>();

        private final Builder() {
        }

        public <T> Builder put(Key<T> key, T value) {
            content.put(key, value);
        }

        public SavedState build() {
            return new SavedState( this );
        }
    }

}
```

Then usage would be like this:

```java
public class LuceneIndexManagerImpl {
    private static SavedState.Key<SavedState> SHARD_HOLDER_KEY = SavedState.key("shard_holder");

    public SavedState saveForRestart() {
        return SavedState.builder()
                .put( SHARD_HOLDER_KEY, shardHolder.save() )
                // ...
                .build();
    }

    public void preStart(SavedState state) {
        shardHolder.preStart( state.get( SHARD_HOLDER_KEY ).orElse( SavedState.EMPTY ) );
    }
}
```

```java
public class Shard {
    private static SavedState.Key<DirectoryHolder> DIRECTORY_HOLDER_KEY = SavedState.key("directory_holder");

    public SavedState saveForRestart() {
        try {
            return SavedState.builder()
                    .put( DIRECTORY_HOLDER_KEY, directoryHolder )
                    .build();
        }
        finally {
            savedForRestart = true;
        }
    }

    public void preStart(SavedState state) {
        directoryHolder = state.get( DIRECTORY_HOLDER_KEY ).orElse( null );
        if ( directoryHolder == null ) {
           // ... create directory holder ...
        }
    }

    public void close() {
        if (!savedForRestart) {
            directoryHolder.close();
        }
    }
}
```

For maps or lists of values, you can probably make do with something like this:

```java
    private static SavedState.Key<Map<Integer, Shard>> SHARDS = SavedState.key("shards");
```

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/directory/impl/LocalHeapDirectoryHolder.java:
```diff
@@ -20,17 +20,36 @@ final class LocalHeapDirectoryHolder implements DirectoryHolder {
 	@Override
 	public void start() {
-		this.directory = new ByteBuffersDirectory( lockFactory );
+		if ( directoryFromPreviousState != null ) {
+			directory = directoryFromPreviousState;
+			return;
+		}
+
+		directory = new ByteBuffersDirectory( lockFactory );
```
I'd just move everything to one method. Call it start or prestart, it shouldn't matter.

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/directory/impl/LocalHeapDirectoryHolder.java:
```diff
@@ -20,17 +20,36 @@ final class LocalHeapDirectoryHolder implements DirectoryHolder {
 
 	private Directory directory;
 
+	private Directory directoryFromPreviousState;
```
You don't really need two separate attributes here.

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/directory/impl/LocalHeapDirectoryHolder.java:
```diff
@@ -20,17 +20,36 @@ final class LocalHeapDirectoryHolder implements DirectoryHolder {
 
+	@Override
+	public void preStart(DirectoryHolder previousDirectoryHolder) {
+		if ( !( previousDirectoryHolder instanceof LocalHeapDirectoryHolder ) ) {
```
That such a thing could even happen seems... concerning? If we support such configuration change, we'll have to test it...

Don't you think it would be simpler to avoid handling state saving in `DirectoryHolder`? We could just reuse the whole `DirectoryHolder` in the shard.

yrodiere commented on engine/src/main/java/org/hibernate/search/engine/common/impl/SearchIntegrationPartialBuildStateImpl.java:
```diff
@@ -164,6 +169,17 @@ public SearchIntegration finalizeIntegration() {
 			}
 			failureCollector.checkNoFailure();
 
+			if ( previousIntegration.isPresent() ) {
```
I'd call `preStart` all the time, and implement `preStart` in such a way that it behaves correctly in either case (reuse saved state if available, create missing resources otherwise). See my other comment about the `SavedState` abstraction.

The reason I would prefer that is to maximize the amount of code common to start and restart. Like I said the other time, the challenge here will be testing; the fewer alternate paths there are to test in the code, the better.

As to what goes in `preStart` and what goes into `start`: in my mind, everything that can be reused after a restart would be initialized in `preStart`, while the rest would be initialized in `start`. But I'm open to discussion about that, of course.

yrodiere reviewed (commented):
I didn't look in depth, but I think we could make this easier to test by using a central abstraction over the saved state and making sure as much code as possible is executed both during normal starts and restarts. See my comments below.

fax4ever commented:
Thanks @yrodiere. I've made the changes.
What else we can improve here?

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/index/impl/ShardingStrategyInitializationContextImpl.java:
```diff
@@ -129,22 +127,18 @@ private void contributeShardWithSilentFailure(Map<String, Shard> shardCollector,
 
-			Shard shard = backendContext.createShard( model, shardEventContext, directoryHolder, ioStrategy,
-					shardPropertySource );
+			Shard shard = backendContext.createShard( model, shardEventContext, directoryProviderHolder.get(), context,
```
Passing and saving the context in the shard/accessor seems very dodgy... most `context` objects are used once and then thrown away, and I'm a bit afraid of what unintended consequences storing a context could have. It could lead to unnecessary memory usage, for example, if the context internally holds a reference to something that is only necessary on startup.

Can't you just delay everything related to the `ioStrategy` (creating the index accessor, ...) until the `preStart` methods are called, just after we create the directory holder? Then you don't need to save the context anymore, you don't need to introduce `start` methods in `IndexReaderProvider`/`IndexWriterProvider` anymore, ...

Delaying all this would make sense regardless of your changes: in a hypothetical Quarkus + Hibernate Search + Lucene extension (or a Quarkus + Infinispan Embedded + Lucene extension), this would allow tuning index writers/readers at runtime, which is definitely something useful.

Also, I thought that was what we mentioned on Zulip initially... ?

If you're uncomfortable with making this change yourself, I can have a look and send a PR, so that you can rebase your changes on that.

yrodiere reviewed (changes requested):
Looks like you're getting close, but I have a few concerns, especially about storing a context object. See my comments below.

fax4ever commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/directory/impl/LocalHeapDirectoryHolder.java:
```diff
@@ -26,7 +26,9 @@ final class LocalHeapDirectoryHolder implements DirectoryHolder {
 	@Override
 	public void start() {
-		this.directory = new ByteBuffersDirectory( lockFactory );
+		if ( directory == null ) {
+			directory = new ByteBuffersDirectory( lockFactory );
+		}
```
if you revert it, you will see that `LuceneIndexRestartFromPreviousIntegrationIT` won't pass anymore.
The check allows us to not create a new Lucene `Directory` when the holder is reused.
I think that actually I should do the same for the `LocalFileSystemDirectoryHolder`.

fax4ever replied:
alternatively we could make the index accessor not call the holder start when the holder is reused, maybe it is cleaner and it would avoid to change the directory holder implementations. Going to use this latter option

fax4ever commented on engine/src/main/java/org/hibernate/search/engine/common/impl/SearchIntegrationPartialBuildStateImpl.java:
```diff
@@ -164,11 +172,30 @@ public SearchIntegration finalizeIntegration() {
+					Map<String, SavedState> indexManagersStates = previousIntegration.get().saveForRestart()
+							.get( INDEX_MANAGERS_KEY ).orElse( Collections.emptyMap() );
+
+					savedState = indexManagersStates.get( entry.getKey() );
```
I think that theoretically there wouldn't be values that are not reused by the new search integration.
But sure, it is better to introduce such a mechanism to prevent any possible leaks in order to support a more generic case.

yrodiere replied:
Yes: it doesn't happen, until it something change in Infinispan for some unrelated feature, and now it does happen, and then we have a leak, and we won't notice it until it's crippling some real-world application... Better safe than sorry.

fax4ever commented on engine/src/main/java/org/hibernate/search/engine/common/impl/SearchIntegrationPartialBuildStateImpl.java:
```diff
@@ -164,11 +172,30 @@ public SearchIntegration finalizeIntegration() {
+					Map<String, SavedState> indexManagersStates = previousIntegration.get().saveForRestart()
+							.get( INDEX_MANAGERS_KEY ).orElse( Collections.emptyMap() );
+
+					savedState = indexManagersStates.get( entry.getKey() );
+				}
+
+				entry.getValue().preStart( failureCollector, beanResolver, propertySource,
+						( savedState == null ) ? SavedState.empty() : savedState );
```
Running the tests, it seems that the null check is still required, since with change the default is applied only if `previousIntegration.isPresent()`. I'm reverting it...

yrodiere replied:
Or you could replace `SavedState savedState = null;` with `SavedState savedState = SavedState.empty();`.

fax4ever replied:
Right, we can init the property using the empty value.

fax4ever commented on engine/src/main/java/org/hibernate/search/engine/common/spi/SearchIntegration.java:
```diff
@@ -21,11 +24,15 @@ public interface SearchIntegration extends AutoCloseable {
 
 	IndexManager indexManager(String indexManagerName);
 
+	SavedState saveForRestart();
```
Actually we need both the saved state and the search integration (that generates it) in the `SearchIntegrationPartialBuildStateImpl`.
We could try something different to not expose this method on SPI, for instance we can have a `SearchIntegrationImpl` instead of `SearchIntegration` as field in `SearchIntegrationPartialBuildStateImpl`.
I'm following the latter solution.

fax4ever commented on backend/lucene/src/test/java/org/hibernate/search/backend/lucene/lowlevel/index/impl/IndexAccessorTest.java:
```diff
@@ -243,9 +256,7 @@ public void computeSizeInBytes_concurrentDeletion() throws IOException {
 
 	private void verifyNoOtherIndexInteractions() {
-		verifyNoMoreInteractions( directoryHolderMock, directoryMock,
-				indexWriterProviderMock, indexWriterDelegatorMock,
-				indexReaderProviderMock, indexReaderMock );
```
With the first commit `HSEARCH-4545 Pass holder to index providers at shard start` the `indexWriterProvider` and `indexReaderProvider` are started ad `IndexAccessorImpl#start`. We removed also the `indexWriterDelegatorMock`, that we didn't need to remove. I'm putting it back...

fax4ever commented:
So far I should have addressed all the comments with exception of the biggest ones:
1. [Closing not used passed states](https://github.com/hibernate/hibernate-search/pull/3029#discussion_r865708829)
2. [Delay ioStrategy actions until the preStart methods are called, just after we create the directory holder](https://github.com/hibernate/hibernate-search/pull/3029#discussion_r865755141)

Tomorrow I'll try to address them.

fax4ever commented:
With the last pushed I tried to address the last comment (2).
In the new solution, the changes on the pre-existing code are minimal and we don't need to store any context in long life objects.

fax4ever commented:
I don't know if it was what you thought.
In any case I still need to address the other comment (1).

yrodiere commented:
> I don't know if it was what you thought.

It's not, though maybe it addresses my concern. I'll have a closer look when you're done and may add a few commits. Please let me know when you're (completely) done.

Thanks.

yrodiere commented:
@fax4ever I created https://hibernate.atlassian.net/browse/HSEARCH-4566 to match more closely what your PR adds. Could you please edit the commit messages to use the new JIRA ticket key?

yrodiere commented:
Thanks. I've started having a look, but I'm not done yet. I'll finish on Monday.

yrodiere reviewed (approved):
I rebased on main, added a few commits and force-pushed. Let's see what CI thinks of this.

fax4ever commented:
> I rebased on main, added a few commits and force-pushed. Let's see what CI thinks of this.

Thank you @yrodiere. I really appreciate the collaboration.

yrodiere commented:
Merged, thanks!
